compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAccessToAcquireRelease.cpp

```diff
@@ -0,0 +1,248 @@
+              AMDAIE::MemoryAccess::Read);
+          rewriter.replaceAllUsesWith(accessOp.getResult(),
+                                      newAccessOp.getResult());
+          logicalObjectFifoToLastAccess[accessOp.getInput()] = accessOp;
+          return WalkResult::advance();
+        });
+    if (res.wasInterrupted()) return failure();
+
+    // Insert release for remaining read access operations at end of block.
+    for (auto &&[value, accessOp] : logicalObjectFifoToLastAccess) {
+      Block *parentBlock = accessOp->getBlock();
+      if (!parentBlock->back().hasTrait<OpTrait::IsTerminator>()) {
+        rewriter.setInsertionPointToEnd(parentBlock);
+      } else {
+        rewriter.setInsertionPoint(parentBlock->getTerminator());
+      }
+      if (!logicalObjectFifoToDma.contains(accessOp.getInput())) {
+        accessOp.emitOpError()
+            << "read access not found as source of DMA operation";
+        return failure();
+      }
+      rewriter.create<AMDAIE::LogicalObjectFifoRelease>(
+          rewriter.getUnknownLoc(), logicalObjectFifoToDma[accessOp.getInput()],
+          LogicalObjectFifoPort::Consume);
+    }
+  }
+  return success();
+}
+
+/// Walk all write access operations within the core operations and insert
+/// semaphore operations. Release operations will be inserted
+/// at the location of the access operation and acquire operations will be
+/// inserted after the preceding access or at the beginning of the block.
+LogicalResult writeAccessToAcquireRelease(Operation *parentOp) {
+  IRRewriter rewriter(parentOp->getContext());
+
+  SmallVector<AMDAIE::CoreOp> coreOps;
+  parentOp->walk([&](AMDAIE::CoreOp coreOp) { coreOps.push_back(coreOp); });
+
+  // Map from DMA source/target logical objectFifos to those respective DMA
+  // operations.
+  DenseMap<Value, AMDAIE::CircularDmaCpyNdOp> logicalObjectFifoToDma;
+  parentOp->walk([&](AMDAIE::CircularDmaCpyNdOp dmaOp) {
+    logicalObjectFifoToDma[dmaOp.getSource()] = dmaOp;
+    logicalObjectFifoToDma[dmaOp.getTarget()] = dmaOp;
+  });
+
+  for (AMDAIE::CoreOp coreOp : coreOps) {
+    DenseMap<Value, SmallVector<AMDAIE::LogicalObjectFifoAccessOp>>
+        logicalObjectFifoToAccesses;
+    DenseMap<Value, AMDAIE::LogicalObjectFifoAccessOp>
+        logicalObjectFifoLastWriteAccesses;
+    WalkResult res = coreOp->walk([&](AMDAIE::LogicalObjectFifoAccessOp
+                                          accessOp) {
+      // If there is another write op on the same logical objectFifo,
+      // release it before the acquire.
+      if (logicalObjectFifoLastWriteAccesses.contains(accessOp.getInput())) {
+        AMDAIE::LogicalObjectFifoAccessOp prevAccess =
+            logicalObjectFifoLastWriteAccesses[accessOp.getInput()];
+        if (!logicalObjectFifoToDma.contains(prevAccess.getInput())) {
+          prevAccess.emitOpError()
+              << "write access not found as source of DMA operation";
+          return WalkResult::interrupt();
+        }
```

Comment:
Why is this needed?

Reply:
For now, the pass is assuming writes are accompanied by DMA operations moving data out of the cores' local memory, and reads are accompanied by DMA operations moving data into local memory.